src/service/git/git-token-service.ts

```diff
@@ -0,0 +1,67 @@
+import { Service } from "typedi";
+
+@Service()
+export class GitTokenService {
+  private githubTokens: Record<string, string>;
+  private gitlabTokens: Record<string, string>;
+  private githubTokenPools: Record<string, string[]>;
+
+  constructor() {
+    this.githubTokens = {};
+    this.gitlabTokens = {};
+    this.githubTokenPools = {};
+  }
+
+  setGithubToken(id: string, token: string) {
+    this.githubTokens[id] = token;
+  }
+
+  setGithubTokenUsingEnv(id: string, tokenId: string) {
+    this.setTokenUsingEnv(id, tokenId, this.githubTokens);
+  }
+
+  getGithubToken(id: string, tokenId: string): string | undefined;
+  getGithubToken(id: string): string | undefined;
+  getGithubToken(id: string, tokenId?: string) {
+    if (tokenId) {
+      this.setGithubTokenUsingEnv(id, tokenId);
+    }
+    return this.githubTokens[id];
+  }
+
+  setGitlabToken(id: string, token: string) {
+    this.gitlabTokens[id] = token;
+  }
+
+  setGitlabTokenUsingEnv(id: string, tokenId: string) {
+    this.setTokenUsingEnv(id, tokenId, this.gitlabTokens);
+  }
+
+  getGitlabToken(id: string, tokenId: string): string | undefined;
+  getGitlabToken(id: string): string | undefined;
+  getGitlabToken(id: string, tokenId?: string) {
+    if (tokenId) {
+      this.setGitlabTokenUsingEnv(id, tokenId);
+    }
+    return this.gitlabTokens[id];
+  }
+
+  setGithubTokenPool(id: string, tokens: string[]) {
+    this.githubTokenPools[id] = tokens;
+  }
+
+  getGithubTokenPool(id: string) {
+    return this.githubTokenPools[id];
+  }
+
+  private setTokenUsingEnv(
+    id: string,
+    tokenId: string,
+    tokenStore: Record<string, string>
+  ) {
+    const token = process.env[tokenId];
+    if (!tokenStore[id] && token) {
```

Comment:
should we really check for this conditional without failing? I'm thinking about a developer wondering why the information is not set on `tokenStore` whenever `setTokenUsingEnv` is called :thinking: 
to set or to fail, right?

Reply:
I decided to add the condition because I was ending up checking this condition anytime I was setting the token from env. So to avoid repetitive code, I decided to put this check in here. If I throw an error if the condition fails then it will defeat the purpose of avoiding repetitive code.

How about I add some JSDoc here so that whenever a developer uses this function they know that it is not going to throw an error?

Reply:
JSDoc is a good idea, please proceed @shubhbapna